src/pendingactions.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module core/pendingactions
+ */
+
+/* global window */
+
+import Plugin from './plugin';
+import ObservableMixin from '@ckeditor/ckeditor5-utils/src/observablemixin';
+import DOMEmitterMixin from '@ckeditor/ckeditor5-utils/src/dom/emittermixin';
+import Collection from '@ckeditor/ckeditor5-utils/src/collection';
+import CKEditorError from '@ckeditor/ckeditor5-utils/src/ckeditorerror';
+
+/**
+ * List of editor pending actions.
+ *
+ * Any asynchronous action that should be finished before the editor destroy (like file upload) should be registered
+ * in this plugin and removed after finish.
+ *
+ * This plugin listens to `window#beforeunload` event and displays browser prompt when a pending action is in progress.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class PendingActions extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'PendingActions';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * DOM Emitter.
+		 *
+		 * @private
+		 * @type {module:utils/dom/emittermixin~EmitterMixin}
+		 */
+		this._domEmitter = Object.create( DOMEmitterMixin );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		/**
+		 * Defines whether there is any registered pending action or not.
+		 *
+		 * @readonly
+		 * @observable
+		 * @type {Boolean} #isPending
+		 */
+		this.set( 'isPending', false );
+
+		/**
+		 * List of pending actions.
+		 *
+		 * @private
+		 * @type {module:utils/collection~Collection}
+		 */
+		this._actions = new Collection( { idProperty: '_id' } );
+		this._actions.delegate( 'add', 'remove' ).to( this );
+
+		// It's not possible to easy test it because karma uses `beforeunload` event
+		// to warn before full page reload and this event cannot be dispatched manually.
+		/* istanbul ignore next */
+		this._domEmitter.listenTo( window, 'beforeunload', ( evtInfo, domEvt ) => {
+			if ( this.isPending ) {
+				domEvt.returnValue = this._actions.get( 0 ).message;
+			}
+		} );
+	}
+
+	/**
+	 * Adds action to the list of pending actions.
+	 *
+	 * Methods returns an object with observable message property. Message can be changed.
+	 *
+	 * @param {String} message
+	 * @returns {Object} Observable object that represents a pending action.
+	 */
+	add( message ) {
+		if ( typeof message !== 'string' ) {
+			throw new CKEditorError( 'pendingactions-add-invalid-message: Message has to be a string.' );
```

Comment:
Missing docs for the `pendingactions-add-invalid-message` error.

See https://github.com/ckeditor/ckeditor5-core/blob/7e7b95007a9f92fe071d9f847536c3499d467f1e/src/plugincollection.js#L186-L197.

Reply:
AFAIK there is no need to add docs to the error when there is no need to provide additional explanation.

Reply:
> we have more errors without docs

Would be good to verify this and report as a new ticket.

Reply:
OTOH I still think that additional explanation for this error is not necessary because the error message is clear and we don't need to explain it in the docs. @pjasiun WDYT? Should we always add docs to error or only when error needs an additional explanation?

Reply:
You don't need to introduce any additional description. See here – https://github.com/ckeditor/ckeditor5-engine/blob/a721f95601511a6fcbb7d43becb957af64f2d157/src/view/attributeelement.js#L109-L117

We need the proper documentation for the error in docs, so

```js
/**
 * Message has to be a string.
 *
 * @error pendingactions-add-invalid-message
 */
```

is enough.

Reply:
I understand. But I'm not sure if documentation for this kind of errors is really necessary. We have more undocumented errors. Let's wait for @pjasiun or @Reinmar comment. If we decided to document all errors because we want to have them in generated docs then it's fine.

Reply:
Ok, I checked and each new type of error needs docs. When the error is repeated then docs is not necessary.